### PR TITLE
Fix for Polygon.normalize() to work on CoordinateSequence (#254)

### PR DESCRIPTION
**Pull request by FObermaier**
* Necessary additions to CoordinateSequences utility class (#188)
  - minCoordinate, minCoordinateIndex
  - indexOf
  - scroll
* Create overload of Orientation.isCCW with CoordinateSequence as ring
  argument partly (#188)

Signed-off-by: Felix Obemaier <felix.obermaier@netcologne.de>

### CODE REVIEW

**Comment by dr-jts** [comment]
This is a fix for #254

File: modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java [hunk 1]
```diff
@@ -220,4 +220,144 @@ public static String toString(CoordinateSequence cs)
...
+   *
+   *@param  seq      the coordinate sequence to rearrange
+   *@param  firstCoordinate  the coordinate to make first
+   */
+  public static void scroll(CoordinateSequence seq, Coordinate firstCoordinate) {
+    int i = indexOf(firstCoordinate, seq);
+    if (i <= 0) return;
+    scroll(seq, i);
+
+    /*
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java]: Should commented code be removed?

File: modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java [hunk 2]
```diff
@@ -220,4 +220,144 @@ public static String toString(CoordinateSequence cs)
...
+  {
+    scroll(seq, indexOfFirstCoordinate, CoordinateSequences.isRing(seq));
+  }
+
+  /**
+   *  Shifts the positions of the coordinates until the coordinate at  <code>firstCoordinateIndex</code>
+   *  is first.
+   *
+   *@param  seq      the coordinate sequence to rearrange
+   *@param  indexOfFirstCoordinate  the index of the coordinate to make first
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java]: Need another param to match signature

File: modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java [hunk 3]
```diff
@@ -220,4 +220,144 @@ public static String toString(CoordinateSequence cs)
...
+   *  The first position is 0; the second, 1; etc.
+   *
+   *@param  coordinate   the <code>Coordinate</code> to search for
+   *@param  seq  the coordinate sequence to search
+   *@return              the position of <code>coordinate</code>, or -1 if it is
+   *      not found
+   */
+  public static int indexOf(Coordinate coordinate, CoordinateSequence seq) {
+    for (int i = 0; i < seq.size(); i++) {
+      if (coordinate.equals(seq.getCoordinate(i))) {
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java]: Better to use equals2D.

Even better to not request a coordinate object...

File: modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java [hunk 4]
```diff
@@ -417,10 +417,17 @@ protected int getSortIndex() {
     return Geometry.SORTINDEX_POLYGON;
   }
 
+  private LinearRing normalized(LinearRing ring, boolean clockwise) {
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java]: Good idea for a better method name.  Might be nice to push this up into Geometry

File: modules/core/src/main/java/org/locationtech/jts/geomgraph/EdgeRing.java [hunk 5]
```diff
@@ -98,7 +98,7 @@ public void computeRing()
       coord[i] = (Coordinate) pts.get(i);
     }
     ring = geometryFactory.createLinearRing(coord);
-    isHole = Orientation.isCCW(ring.getCoordinates());
+    isHole = Orientation.isCCW(ring.getCoordinateSequence());
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geomgraph/EdgeRing.java]: Perhaps this should be a separate PR, since I don't think it is related to normalize issue?

File: modules/core/src/main/java/org/locationtech/jts/operation/polygonize/EdgeRing.java [hunk 6]
```diff
@@ -209,7 +209,7 @@ public boolean isHole()
   public void computeHole()
   {
     LinearRing ring = getRing();
-    isHole = Orientation.isCCW(ring.getCoordinates());
+    isHole = Orientation.isCCW(ring.getCoordinateSequence());
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/operation/polygonize/EdgeRing.java]: Perhaps better in a separate PR?

File: modules/core/src/test/java/org/locationtech/jts/triangulate/ConformingDelaunayTest.java [hunk 7]
```diff
@@ -71,6 +71,7 @@ void runDelaunay(String sitesWKT, String constraintsWKT, boolean computeTriangle
   	Geometry expectedEdges = reader.read(expectedWKT);
   	result.normalize();
   	expectedEdges.normalize();
-  	assertTrue(expectedEdges.equalsExact(result, COMPARISON_TOLERANCE));
+  	assertTrue("Assertion failed!\nexpected: " + expectedEdges.toString() +"\result:    " + result.toString(),
```
Comment (dr-jts) [inline modules/core/src/test/java/org/locationtech/jts/triangulate/ConformingDelaunayTest.java]: Should not be part of this PR

File: modules/core/src/test/java/org/locationtech/jts/triangulate/VoronoiTest.java [hunk 8]
```diff
@@ -66,6 +66,7 @@ void runVoronoi(String sitesWKT, boolean computeTriangles, String expectedWKT)
   	Geometry expectedEdges = reader.read(expectedWKT);
   	result.normalize();
   	expectedEdges.normalize();
-  	assertTrue(expectedEdges.equalsExact(result, COMPARISON_TOLERANCE));
+  	assertTrue("Assertion failed!\nexpected: " + expectedEdges.toString() +"\nresult:   " + result.toString(),
```
Comment (dr-jts) [inline modules/core/src/test/java/org/locationtech/jts/triangulate/VoronoiTest.java]: Should not be part of this PR

File: modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java [hunk 9]
```diff
@@ -220,4 +220,144 @@ public static String toString(CoordinateSequence cs)
...
\ No newline at end of file
+
+  /**
+   *  Returns the minimum coordinate, using the usual lexicographic comparison.
+   *
+   *@param  seq  the coordinate sequence to search
+   *@return  the minimum coordinate in the sequence, found using <code>compareTo</code>
+   *@see Coordinate#compareTo(Object)
+   */
+  public static Coordinate minCoordinate(CoordinateSequence seq)
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequences.java]: Ideally will have separate unit tests for this code

File: modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java (hunk 4 above)
Comment (FObermaier) [reply modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java]: This would be another PR, wouldn't it?

File: modules/core/src/main/java/org/locationtech/jts/geomgraph/EdgeRing.java (hunk 5 above)
Comment (FObermaier) [reply modules/core/src/main/java/org/locationtech/jts/geomgraph/EdgeRing.java]: Agreed, my enthusiasm. Belongs to #188.

File: modules/core/src/main/java/org/locationtech/jts/operation/polygonize/EdgeRing.java (hunk 6 above)
Comment (FObermaier) [reply modules/core/src/main/java/org/locationtech/jts/operation/polygonize/EdgeRing.java]: dito

File: modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java (hunk 4 above)
Comment (dr-jts) [reply modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java]: Yes, so not needed for this PR

File: modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java [hunk 10]
```diff
@@ -417,10 +417,17 @@ protected int getSortIndex() {
...
+    LinearRing res = getFactory().createLinearRing(ring.getCoordinateSequence().copy());
+    normalize(res, clockwise);
+    return res;
+  }
+
   private void normalize(LinearRing ring, boolean clockwise) {
     if (ring.isEmpty()) {
       return;
     }
+    /* old code
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java]: Should delete old code

File: modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java (hunk 4 above)
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/Polygon.java]: Would be better to move this to LinearRing class

File: modules/core/src/test/java/org/locationtech/jts/geom/CoordinateSequencesTest.java [hunk 11]
```diff
@@ -67,4 +125,259 @@ private static CoordinateSequence createTestSequence(CoordinateSequenceFactory c
     return cs;
   }
 
+  private static CoordinateSequence createRandomTestSequence(CoordinateSequenceFactory csFactory, int size, int dim,
```
Comment (dr-jts) [inline modules/core/src/test/java/org/locationtech/jts/geom/CoordinateSequencesTest.java]: It's maybe a style thing, but I prefer to not use randomly generated data in unit tests. It makes it hard to reproduce behaviour.  (Its fine for performance testing, though - but JTS has a different set of packages for this).

**Comment by dr-jts** [comment]
Thx for doing all this work, @FObermaier.    Hopefully my review comments help to make the JTS codebase better...

File: modules/core/src/test/java/org/locationtech/jts/geom/CoordinateSequencesTest.java (hunk 11 above)
Comment (dsmiley) [reply modules/core/src/test/java/org/locationtech/jts/geom/CoordinateSequencesTest.java]: FYI https://labs.carrotsearch.com/randomizedtesting.html is some cool test tooling to help with randomizations -- most importantly to make a randomized test _easily repeatable_.  Lucene/Solr uses this extensively and I've become a big proponent of the methodology.

If randomized tests aren't acceptable in JTS, I suggest @FObermaier merely tweak this test he already invested time in to simply use a constant seed -- `new Random(0)`.
